rllib/agents/qmix/model.py

```diff
@@ -31,7 +31,7 @@ def get_initial_state(self):
 
     @override(TorchModelV2)
     def forward(self, input_dict, hidden_state, seq_lens):
-        x = F.relu(self.fc1(input_dict["obs_flat"].float()))
+        x = F.relu(self.fc1(input_dict["obs"].float()))
```

Comment:
Would it also work to be obs_flat? That would let it handle complex observation spaces which wouldn't be a flat vector.

Reply:
Unless I am mistaken, obs_flat includes all of the elements in the agent's obs dict, flattened and concatted, which, if we are putting the environment global state in the agents' obs dict, means that the agent's policy gets that info. We could make it so there is an added env state obs dict alongside the grouped agent obs dicts that come out of GroupAgentsWrapper (i.e., the env's observation space is a (potentially nested) dict with keys eg "env_state", "group_1", ...). That would require some refactoring of that wrapper and probably some other parts of the code.

Reply:
Actually, I just realized that if you look here: https://github.com/ray-project/ray/blob/785670bc18a8595219c96e9512192922fafcf510/rllib/agents/qmix/qmix_policy.py#L426-L443

, which is the only part of qmix_policy.py where the NN's forward pass is called, there is an assumption here, before forward() is entered, that the observation tensor for each agent is a vector.

Reply:
I think obs flat and obs are equivalent then, since you pass in only the agent obs to model()? It should not be able to find the original obs that has the env state.

Reply:
I think that is correct here. It looks like it was done this way in the original writing of qmix_policy where a new dict that is just the "obs" tensor is passed into the NN model so that the "action_mask" tensor would not be passed in in obs_flat.